module_2/gift_certificates_system/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,254 @@
+package com.epam.esm.service.impl;
+
+import com.epam.esm.dao.impl.GiftCertificateDaoImpl;
+import com.epam.esm.dao.impl.GiftCertificateToTagRelationDaoImpl;
+import com.epam.esm.dao.impl.TagDaoImpl;
+import com.epam.esm.entity.GiftCertificate;
+import com.epam.esm.entity.Tag;
+import com.epam.esm.exception.EntityExistsException;
+import com.epam.esm.exception.EntityNotFoundException;
+import com.epam.esm.service.GitCertificateService;
+import com.epam.esm.util.ColumnName;
+import com.epam.esm.util.SortOperator;
+import lombok.extern.log4j.Log4j2;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+import java.lang.reflect.Field;
+import java.time.LocalDateTime;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * Gift certificate service implementation.
+ */
+@Log4j2
+@Service
+public class GiftCertificateServiceImpl extends GitCertificateService {
+    private final GiftCertificateToTagRelationDaoImpl relationDao;
+
+    /**
+     * Instantiates a new Gift certificate service.
+     *
+     * @param giftCertificateDao    - Gift certificate DAO layer.
+     * @param tagDao                - Tag DAO layer.
+     * @param relationDao           - Gift certificate to tag relation DAO layer.
+     */
+    @Autowired
+    public GiftCertificateServiceImpl(GiftCertificateDaoImpl giftCertificateDao, TagDaoImpl tagDao, GiftCertificateToTagRelationDaoImpl relationDao) {
+        super(giftCertificateDao, tagDao);
+        this.relationDao = relationDao;
+    }
+
+    @Transactional
+    @Override
+    public GiftCertificate create(GiftCertificate entity) {
+        String name = entity.getName();
+        Optional<GiftCertificate> optionalGiftCertificate = giftCertificateDao.findByName(name);
+
+        if(optionalGiftCertificate.isPresent()) {
+            log.debug("Gift certificate with name '{}' already exists in the database.", entity.getName());
+            throw new EntityExistsException("Gift certificate with name '" + entity.getName() + "' already exists in the database.");
```

Comment:
Please don't log and throw exception at the same time

Reply:
Thanks, fixed.